apps/wallet/src/shared/qredo-api.ts

```diff
@@ -0,0 +1,155 @@
+    apiData: QredoAPIErrorResponse;
+
+    constructor(status: number, apiData: QredoAPIErrorResponse) {
+        super(`Qredo API Error (status: ${status}). ${apiData.msg}`);
+        this.status = status;
+        this.apiData = apiData;
+    }
+}
+
+export class QredoAPIUnauthorizedError extends QredoAPIError {}
+
+export type AccessTokenParams = {
+    refreshToken: string;
+    grantType?: string;
+};
+
+export type AccessTokenResponse = {
+    access_token: string;
+    expires_in: number;
+    token_type: string;
+};
+
+export type Wallet = {
+    walletID: string;
+    address: string;
+    network: string;
+    labels: {
+        key: string;
+        name: string;
+        value: string;
+    }[];
+};
+
+export type GetWalletsResponse = {
+    wallets: Wallet[];
+};
+
+export type GetWalletsParams = {
+    filters?: { address?: string };
+};
+
+export type AccessTokenRenewalFunction = (
+    qredoID: string
+) => Promise<string | null>;
+
+export class QredoAPI {
+    readonly baseURL: string;
+    readonly qredoID: string;
+    #accessToken: string | null;
+    #renewAccessTokenFN: AccessTokenRenewalFunction | null;
+    #accessTokenRenewInProgress: ReturnType<AccessTokenRenewalFunction> | null =
+        null;
+
+    constructor(
+        qredoID: string,
+        baseURL: string,
+        options: {
+            accessToken?: string;
+            accessTokenRenewalFN?: AccessTokenRenewalFunction;
+        } = {}
+    ) {
+        this.qredoID = qredoID;
+        this.baseURL = baseURL + (baseURL.endsWith('/') ? '' : '/');
+        this.#accessToken = options.accessToken || null;
+        this.#renewAccessTokenFN = options.accessTokenRenewalFN || null;
+    }
+
+    public set accessToken(accessToken: string) {
+        this.#accessToken = accessToken;
+    }
+
+    public get accessToken() {
+        return this.#accessToken || '';
+    }
+
+    public createAccessToken({
+        refreshToken,
+        grantType = 'refresh_token',
+    }: AccessTokenParams): Promise<AccessTokenResponse> {
+        const params = new FormData();
+        params.append('refresh_token', refreshToken);
+        if (grantType) {
+            params.append('grant_type', grantType);
+        }
+        return this.#request(`${this.baseURL}connect/sui/token`, {
+            method: 'post',
+            body: params,
+        });
+    }
+
+    public getWallets({
+        filters,
+    }: GetWalletsParams = {}): Promise<GetWalletsResponse> {
+        const searchParams = new URLSearchParams();
+        if (filters?.address) {
+            searchParams.append('address', filters.address);
+        }
+        const searchQuery = searchParams.toString();
+        return this.#request(
+            `${this.baseURL}connect/sui/wallets${
+                searchQuery ? `?${searchQuery}` : ''
+            }`
+        );
+    }
+
+    #request = async (...params: Parameters<typeof fetch>) => {
+        let tries = 0;
+        while (tries++ <= 1) {
```

Comment:
nit: it might be nice to make `maxRetries` an optional arg that defaults to 1 so we don't have to reference `1` in multiple places which is slightly error prone